src/piadm.sh

```diff
@@ -0,0 +1,741 @@
+	if [[ "$iso" == "yes" ]]; then
+	    umount ${tdir}/mnt
+	fi
+	/bin/rm -rf ${tdir}
+	echo "PI-stamp $stamp appears to be already on /${bootfs}"
+	err "Use   piadm remove $stamp   to remove any old copies."
+    fi
+    mkdir /${bootfs}/platform-${stamp}
+    tar -cf - -C ${tdir}/mnt/platform . | \
+	tar -xf - -C /${bootfs}/platform-${stamp}
+
+    if [[ "$iso" == "yes" ]]; then
+	umount ${tdir}/mnt
+    fi
+    /bin/rm -rf ${tdir}
+
+    # Global variable for enablepool() usage...
+    installstamp=$stamp
+    return 0
+}
+
+list() {
+    if [[ $1 == "-H" ]]; then
+	pool=$2
+    else
+	printf "%-18s %-30s %-12s %-5s %-5s \n" "PI STAMP" \
+	   "BOOTABLE FILESYSTEM" "BOOT BITS?" "NOW" "NEXT"
+	pool=$1
+    fi
+
+    poolpresent $pool
+
+    getbootable
+    for bootfs in ${allbootfs[@]}; do
+	bfspool=$(echo $bootfs | awk -F/ '{print $1}')
+	if [[ "$pool" != "" && "$bfspool" != "$pool" ]]; then
+	    # If we specify a pool for listing, skip ones not in the pool.
+	    continue
+	fi
+	if [[ ! -L /$bootfs/platform ]]; then
+	    corrupt \
+		"WARNING: Bootable filesystem $bootfs has non-symlink platform"
+	fi
+	cd /$bootfs
+	bootbitsstamp=$(cat etc/version/boot)
+	bootstamp=$(cat platform/etc/version/platform)
+	mapfile -t pis \
+		< <(cd /$bootfs ; cat platform-*/etc/version/platform)
+	for pi in ${pis[@]}; do
+	    if [[ $activestamp == $pi ]]; then
+		active="yes"
+	    else
+		active="no"
+	    fi
+	    if [[ $bootstamp == $pi ]]; then
+		booting="yes"
+	    else
+		booting="no"
+	    fi
+	    if [[ $bootbitsstamp == $pi ]]; then
+		bootbits="next"
+	    elif [[ -d boot-$pi ]]; then
+		bootbits="available"
+	    else
+		bootbits="none"
+	    fi
+	    printf "%-18s %-30s %-12s %-5s %-5s\n" $pi $bootfs $bootbits \
+		$active $booting
+	done
+    done
+}
+
+update_boot_sectors() {
+    pool=$1
+    bootfs=$2
+
+    # XXX WARNING -- illumos#12894 will allow slogs.  We will need to
+    # alter the generation of boot_devices accordingly.
+    # Generate the pool's boot devices now, in case we did something
+    # hyper-clever for the pool.  s1 may be created, but not yet PCFS...
+    mapfile -t boot_devices < <(zpool list -v "$pool" | \
+        grep -E 'c[0-9]+' | awk '{print $1}' | sed -E 's/s[0-9]+//g')
+
+    # Reality check the pool was created with -B.
+    # First way to do this is to check for the `bootsize` property not
+    # its default, which is NO bootsize.
+    zpool get bootsize $pool | grep -q -w default
+    if [[ $? -eq 0 ]]; then
+	# No bootsize is a first-cut test.  It passes if the pool was
+	# created with `zpool create -B`. There's one other that needs
+	# to be peformed, because some bootable pools are manually
+	# configured to share slices with other functions (slog,
+	# l2arc, dedup):
+
+	# Use fstyp to confirm if this is a stealth EFI booter...
+	type=$(fstyp /dev/dsk/${boot_devices[0]}s0)
+	if [[ "$type" == "pcfs" ]]; then
+	    # If we detect PCFS on s0, it's LIKELY an EFI System Partition that
```

Comment:
do we even need or want to support a pool without "bootsize" ?

Reply:
Yes, we do, because of BIOS-era pools.

```
[root@moe (kebecloud) ~]# piadm list
PI STAMP           BOOTABLE FILESYSTEM            BOOT BITS?   NOW   NEXT  
20200728T010227Z   zones/boot                     next         yes   yes  
[root@moe (kebecloud) ~]# piadm bootable
zones                          ==> BIOS 
[root@moe (kebecloud) ~]# zpool get bootsize
NAME   PROPERTY  VALUE     SOURCE
zones  bootsize  -         default
[root@moe (kebecloud) ~]# 
```

Reply:
My corner case here is the "double-duty ssd" case where it's sliced into pieces for boot and slog-or-l2arc.